File: apps/web/lib/grounded.ts
/**
 * Evidence-grounded answering client.
 *
 * Requests are made by the Next.js server runtime, matching the rest of the app:
 * the API needs no browser exposure, and a question about unpublished patent
 * work never leaves the server it was typed into.
 *
 * Note what this module does *not* export: anything that would let a caller
 * choose a model, a provider, a temperature, or a prompt. The API rejects those
 * fields outright, and offering them here would only produce a 422.
 */

import { API_BASE_URL } from "@/lib/api";
import type { ClaimType } from "@/lib/claims";
import type { SourceLocator } from "@/lib/documents";
import type { RetrievalMode, RetrievalProfile } from "@/lib/search";

/** Mirrors the API's bounds. The server enforces them; these keep the UI honest. */
export const MAX_QUESTION_LENGTH = 512;
export const GROUNDED_TOP_K_CHOICES = [3, 5, 6, 10] as const;

export type InsufficientReason =
  | "no_retrieved_evidence"
  | "evidence_not_specific_enough"
  | "conflicting_evidence"
  | "question_outside_available_documents";

/** One canonical span, with the text the server read at those exact offsets. */
export type GroundedSourceSpan = {
  locator: SourceLocator;
  /**
   * Resolved from stored page text by the server, never reproduced by the
   * model. Render it as text.
   */
  quote: string;
};

export type GroundedEvidence = {
  /** Issued for this request only. Meaningless in any other. */
  evidence_id: string;
  document_id: string;
  document_name: string;
  claim_number: number;
  claim_type: ClaimType;
  depends_on: number[];
  source_spans: GroundedSourceSpan[];
  crosses_pages: boolean;
  fused_rank: number;
  fused_score: number;
  /** Null means the dense channel did not retrieve this claim. Never render 0. */
  dense_rank: number | null;
  dense_score: number | null;
  lexical_rank: number | null;
  lexical_score: number | null;
};

export type GroundedStatement = {
  text: string;
  /** Never empty, and every entry appears in `evidence`. */
  evidence_ids: string[];
};

export type GroundedRetrieval = {
  mode: RetrievalMode;
  profile: RetrievalProfile;
  searched_index_run_count: number;
  retrieved_candidate_count: number;
  included_evidence_count: number;
  omitted_evidence_count: number;
};

/**
 * Token counts, each independently nullable.
 *
 * "Not reported" and "zero" are different facts. A UI that renders a fabricated
 * 0 as a measurement is worse than one that renders a dash.
 */
export type GroundedUsage = {
  input_tokens: number | null;
  output_tokens: number | null;
  total_tokens: number | null;
};

export type GroundedGenerationMetadata = {
  provider: string;
  model: string;
  model_version: string | null;
  finish_reason: string;
  usage: GroundedUsage;
  duration_seconds: number;
  attempts: number;
  structured_output_mode: string | null;
  warnings: string[];
};

export type GroundedAnswer = {
  /** Composed by the server from validated statements. Plain text, never HTML. */
  answer: string;
  statements: GroundedStatement[];
  evidence: GroundedEvidence[];
  /** True is a normal result, not an error. */
  insufficient_evidence: boolean;
  insufficient_reason: InsufficientReason | null;
  retrieval: GroundedRetrieval;
  /** Null when retrieval returned nothing and no provider was contacted. */
  generation: GroundedGenerationMetadata | null;
  warnings: string[];
};

export type GroundedAnswerRequest = {
  query: string;
  document_ids?: string[];
  mode?: RetrievalMode;
  top_k?: number;
};

export type GroundedOutcome =
  | { ok: true; answer: GroundedAnswer }
  | { ok: false; detail: string; errorCode: string };

/**
 * Ask a grounded question.
 *
 * Rejections are returned rather than thrown: an unreachable model server and a
 * fabricated citation are both expected outcomes the page has to render, not
 * crashes.
 */
export async function askGrounded(
  request: GroundedAnswerRequest,
): Promise<GroundedOutcome> {
  let response: Response;
  try {
    response = await fetch(`${API_BASE_URL}/api/v1/grounded/answers`, {
      method: "POST",
      headers: { "Content-Type": "application/json", Accept: "application/json" },
      body: JSON.stringify(request),
      cache: "no-store",
    });
  } catch {
    return {
      ok: false,
      detail: "Could not reach the API. Check that the backend is running.",
      errorCode: "api_unreachable",
    };
  }

  const payload: unknown = await response.json().catch(() => null);

  if (response.ok) {
    return { ok: true, answer: payload as GroundedAnswer };
  }

  const error = (payload ?? {}) as { detail?: string; error_code?: string };
  return {
    ok: false,
    detail: error.detail ?? "The question could not be answered.",
    errorCode: error.error_code ?? String(response.status),
  };
}

/** Operator-facing wording for the closed set of limitation reasons. */
export function insufficientReasonLabel(reason: InsufficientReason | null): string {
  switch (reason) {
    case "no_retrieved_evidence":
      return "Nothing retrieved";
    case "evidence_not_specific_enough":
      return "Evidence not specific enough";
    case "conflicting_evidence":
      return "Conflicting evidence";
    case "question_outside_available_documents":
      return "Outside the indexed documents";
    default:
      return "Insufficient evidence";
  }
}

/**
 * Look one evidence entry up by the id a statement cited.
 *
 * Returns undefined rather than throwing. The API guarantees every cited id is
 * present, but a UI that crashes when a guarantee is broken is worse at telling
 * you the guarantee was broken.
 */
export function findEvidence(
  evidence: GroundedEvidence[],
  evidenceId: string,
): GroundedEvidence | undefined {
  return evidence.find((entry) => entry.evidence_id === evidenceId);
}

/** "3 of 7 retrieved claims were used as evidence" — what the reader is looking at. */
export function evidenceSummary(retrieval: GroundedRetrieval): string {
  const parts = [
    `${retrieval.included_evidence_count} of ${retrieval.retrieved_candidate_count} retrieved claim(s) given to the model`,
  ];
  if (retrieval.omitted_evidence_count > 0) {
    parts.push(`${retrieval.omitted_evidence_count} omitted for context budget`);
  }
  return parts.join(" · ");
}

File: apps/api/src/claimtrace_api/schemas/grounded.py
"""Request and response models for evidence-grounded answering.

The request model is mostly interesting for what it refuses. ``extra="forbid"``
is not tidiness here: it is the enforcement point for the rule that a caller
controls the *question*, never the machinery. A body carrying ``model``,
``provider``, ``system``, ``temperature``, ``seed``, ``schema``,
``evidence_ids``, ``page_number``, or ``context`` is rejected with a 422 rather
than silently ignored - and silently ignoring it would be worse, because a
client could ship code that appears to pin a model for a year before anyone
notices it never did.

The response is entirely server-owned. Every field in it is either something the
server computed, something it retrieved, or something a model said that survived
citation validation. There is no field into which unvalidated model text can
reach a reader.
"""

from __future__ import annotations

import uuid

from pydantic import BaseModel, ConfigDict, Field, field_validator

from claimtrace_api.db.models import ClaimType
from claimtrace_api.grounding.draft import InsufficientReason
from claimtrace_api.retrieval.base import RetrievalMode
from claimtrace_api.schemas.llm import GenerationMetadataResponse
from claimtrace_api.schemas.locators import SourceLocator
from claimtrace_api.schemas.retrieval import (
    MAX_DOCUMENT_FILTER,
    MAX_QUERY_LENGTH,
    MAX_TOP_K,
    RetrievalProfileResponse,
)


class GroundedAnswerRequest(BaseModel):
    """One grounded question.

    Deliberately the same four knobs a claim search takes, minus the two
    candidate counts. Those are a cost decision rather than a relevance one, and
    a caller that could inflate them could make one request sweep the whole
    index; the server picks them from configuration.
    """

    model_config = ConfigDict(extra="forbid")

    query: str = Field(min_length=1, max_length=MAX_QUERY_LENGTH)
    #: Absent or empty means every indexed document. Duplicates are collapsed.
    document_ids: list[uuid.UUID] = Field(default_factory=list, max_length=MAX_DOCUMENT_FILTER)
    mode: RetrievalMode = RetrievalMode.HYBRID
    top_k: int = Field(default=6, ge=1, le=MAX_TOP_K)

    @field_validator("query")
    @classmethod
    def _reject_blank(cls, value: str) -> str:
        """A whitespace-only question passes min_length and retrieves nothing."""
        if not value.strip():
            raise ValueError("query must not be blank")
        return value


class GroundedSourceSpanResponse(BaseModel):
    """One canonical span, with the text stored at those coordinates.

    ``quote`` is read out of ``document_pages.text`` at ``locator``'s offsets by
    the server. It is never a model's reproduction of the text, which would be a
    second assertion about the source rather than a consequence of it.
    """

    locator: SourceLocator
    quote: str


class GroundedEvidenceResponse(BaseModel):
    """One piece of cited evidence, as the reader sees it.

    Carries the retrieval ranks so a reader can tell *why* this claim was in
    front of the model. Absence is preserved: a null ``dense_rank`` means the
    dense channel did not retrieve this claim, and must not be rendered as zero.
    """

    #: The identifier issued for this request. Meaningless in any other request,
    #: and returned only so statements can be matched to their evidence.
    evidence_id: str
    document_id: uuid.UUID
    document_name: str
    claim_number: int
    claim_type: ClaimType
    depends_on: list[int] = Field(default_factory=list)
    #: In span order. A claim crossing a page break keeps one span per page.
    source_spans: list[GroundedSourceSpanResponse]
    crosses_pages: bool

    fused_rank: int
    fused_score: float
    dense_rank: int | None = None
    dense_score: float | None = None
    lexical_rank: int | None = None
    lexical_score: float | None = None


class GroundedStatementResponse(BaseModel):
    """One statement and the evidence it was validated against."""

    text: str
    #: Never empty, deduplicated, and every entry present in ``evidence``.
    evidence_ids: list[str]


class GroundedRetrievalResponse(BaseModel):
    """What was searched, and how much of it reached the model."""

    mode: RetrievalMode
    profile: RetrievalProfileResponse
    #: Completed index runs the active profile matched. Zero means nothing is
    #: indexed for this profile - distinct from a question that matched nothing.
    searched_index_run_count: int
    retrieved_candidate_count: int
    included_evidence_count: int
    #: Retrieved claims dropped to stay inside the context budget. These were
    #: never shown to the model and could not have been cited.
    omitted_evidence_count: int


class GroundedAnswerResponse(BaseModel):
    """A grounded answer, or an honest account of why there is not one."""

    #: Composed by the server from the validated statements below, plus a fixed
    #: limitation sentence when evidence was insufficient. Plain text: render it
    #: as text, never as HTML or Markdown.
    answer: str
    statements: list[GroundedStatementResponse] = Field(default_factory=list)
    #: Only evidence actually cited by a validated statement. Retrieved claims
    #: the model did not use are not returned as though they supported anything.
    evidence: list[GroundedEvidenceResponse] = Field(default_factory=list)

    #: True is a normal 200 outcome, not an error. It means the retrieved claims
    #: do not answer the question - which is the correct answer to a great many
    #: questions, and the one a system like this must be willing to give.
    insufficient_evidence: bool
    insufficient_reason: InsufficientReason | None = None

    retrieval: GroundedRetrievalResponse
    #: ``None`` when no provider was contacted, which happens when retrieval
    #: returned nothing at all.
    generation: GenerationMetadataResponse | None = None
    warnings: list[str] = Field(default_factory=list)
